thunder/core/transforms.py

```diff
@@ -3768,9 +3769,91 @@ def backward_fn(saved_for_backward, cotangents):
+    """Generates the pair of traces with rematerializaion of the saved-for-backward tensors.
+    Args:
+        fw_trace (Trace): forward trace where to get the saved for backward from.
+        bw_trace (Trace): backward trace where to recompute the saved for backward to.
+
+    Returns:
+        tuple[Trace, Trace]: A tuple containing the new forward and backward traces.
+    """
+    start_time_ns = time.perf_counter_ns()
+    # First of all we need to get the saved for backward tensors
+    saved_for_bw = get_saved_for_backward_tensors(fw_trace)
+
+    # Let's create these two maps that will help writing the logic
+    fw_trace_args = {j.name: j for j in fw_trace.args}
+    old_saved_for_bw_map = {j.name: j for j in saved_for_bw}
+
+    # The second step is to determine which tensors we can rematerialize. Let's remove the tensors that come from the argument to the forward tarce.
+    recomputable_map = {
+        old_saved_for_bw_map[i].name: old_saved_for_bw_map[i] for i in set(old_saved_for_bw_map) - set(fw_trace_args)
+    }
+
+    # Here will come the selection logic for what to materialize. At the moment let's remat everything and just print some stats.
+    from collections import Counter
+
+    tensor_counter = Counter({k: v.numel * v.dtype.bytes for k, v in recomputable_map.items()})
+    print(f"DEBUG: out of {len(saved_for_bw)} sfb, {len(recomputable_map)} are recomputable")
+    print(f"DEBUG: can save up to {tensor_counter.total()*1e-6:.3f} MB")
+
+    # Now that we have the list of tensors that are going to be recomputed we can start to collect the computation needed to generate them.
+    prod_symbols = find_producer_symbols(fw_trace, recomputable_map.values(), fw_trace.args)
+    # We need to check which inputs to the fw trace are used to compute the producer symbols
+    required_fw_args = {}
+    for p in prod_symbols:
+        for pa in p.flat_args:
+            if isinstance(pa, TensorProxy) and pa.name in fw_trace_args.keys():
+                required_fw_args[pa.name] = pa
+
+    # Now that everything is ready we can prepare the new saved for backward list
+    # To do so we need: the saved for backward that still needs to be carried on from the old list and the new required arguments from the forward.
+    required_saved_for_bw = {
+        fw_trace_args[i].name: fw_trace_args[i] for i in set(fw_trace_args) & set(old_saved_for_bw_map)
+    }
+    new_saved_for_backward = tuple((required_saved_for_bw | required_fw_args).values())
+
+    # Everything is now ready and we can start to build the new traces.
+    new_fw_trace = from_trace(fw_trace)
+    new_fw_trace.bound_symbols = fw_trace.bound_symbols.copy()
+    new_return = (fw_trace.output[0], (new_saved_for_backward, fw_trace.output[1][1]))
+    new_fw_trace.bound_symbols[-1] = prims.python_return.bind(new_return, output=())
+
+    new_bw_trace = from_trace(bw_trace)
+    new_bsym = bw_trace.bound_symbols.copy()
+
+    with tracectx(new_bw_trace):
```

Comment:
Do we need `tracectx`, as we are explicitly passing the `name="C0"`?

Reply:
Unfortunately not because the proxy creation ends up checking if the trace contains the same name already here:

https://github.com/Lightning-AI/lightning-thunder/blob/34fed98e71957afb215cc1c5be6c9b3924e82d19/thunder/core/proxies.py#L66-L68